component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/EventHubClient.java

```diff
@@ -32,74 +45,218 @@
+            log.debug("Unrecoverable error occurred when creating Eventhub Client using following attributes. Auth "
+                              + "endpoint: " + authEndpoint.replaceAll("[\r\n]", "") + ". Analytics event publishing "
+                              + "will be disabled until issue is rectified. Reason: "
+                              + e.getMessage().replaceAll("[\r\n]", ""), e);
         }
-        batch = producer.createBatch();
-        readWriteLock = new ReentrantReadWriteLock();
-        sendSemaphore = new Semaphore(1);
     }
 
     public void sendEvent(String event) {
-        if (producer == null) {
-            log.error("EventHubClient has failed. Hence ignoring.");
-            return;
-        }
-        EventData eventData = new EventData(event);
-        readWriteLock.readLock().lock();
-        try {
-            boolean isAdded = batch.tryAdd(eventData);
-            if (!isAdded) {
-                try {
-                    sendSemaphore.acquire();
-                    isAdded = batch.tryAdd(eventData);
-                    if (!isAdded) {
-                        int size = batch.getCount();
-                        producer.send(batch);
-                        batch = producer.createBatch();
+        if (clientStatus == ClientStatus.CONNECTED) {
+            EventData eventData = new EventData(event);
+            readWriteLock.readLock().lock();
+            try {
+                boolean isAdded = batch.tryAdd(eventData);
+                if (!isAdded) {
+                    try {
+                        readWriteLock.readLock().unlock();
+                        readWriteLock.writeLock().lock();
                         isAdded = batch.tryAdd(eventData);
-                        log.debug("Published " + size + " events to Analytics cluster.");
+                        if (!isAdded) {
+                            int size = batch.getCount();
+                            producer.send(batch);
+                            batch = createBatchWithRetry();
+                            isAdded = batch.tryAdd(eventData);
+                            log.info("Published " + size + " events to Analytics cluster.");
+                        }
+                    } catch (Exception e) {
+                        if (e.getCause() instanceof TimeoutException) {
+                            log.error("Timeout occurred after retrying " + retryOptions.getMaxRetries() + " "
+                                              + "times with an timeout of " + retryOptions.getTryTimeout() + " "
+                                              + "seconds while trying to publish EventDataBatch. Next retry cycle "
+                                              + "will begin shortly.");
+                            readWriteLock.writeLock().unlock();
+                            sendEvent(event);
+                        } else if (e instanceof AmqpException && isAuthenticationFailure((AmqpException) e)) {
```

Comment:
Rather than catching Exception class and then checking for each type,can we catch each specific Exception and then use other methods to divide it into more specific scenarios?

Reply:
Fixed with 6f9169c5444fc677f7378b5e5b0d9b0a0d5fd920